src/mug/ssl.gleam

```diff
@@ -0,0 +1,334 @@
+import gleam/dynamic.{type Dynamic}
+import gleam/erlang/atom
+import gleam/erlang/charlist.{type Charlist}
+import gleam/erlang/process
+import gleam/io
+import gleam/result
+import mug.{type ConnectionOptions, type Error}
+
+pub type Socket
+
+type DoNotLeak
+
+pub type SSLConnectionOptions {
+  SSLConnectionOptions(
+    host: String,
+    port: Int,
+    timeout: Int,
+    cacerts: CaCertificates,
+  )
+}
+
+/// The CA certificates to use
+pub type CaCertificates {
+  /// Do not verify certificates.
+  NoVerification
+  /// Use these der-encoded certificates as CA certificates. This will make mug not use
+  /// the system's CA certificates, but only these ones. To use the system's certificates
+  /// and custom ones, use the `WithSystemCertificatesAnd` variant instead.
+  CustomDerCertificates(cacerts: List(BitArray))
+  /// Path to a pem-encoded file which contains CA certificates. This will make mug not use
+  /// the system's CA certificates, but only these ones. To use the system's certificates
+  /// and custom ones, use the `WithSystemCertificatesAnd` variant instead.
+  CustomPemCertificates(cacertfile: String)
+  /// Use the system's CA certificates, as provided by erlang's `cacerts_get/0` function
+  /// from the `public_key` module ([docs](https://www.erlang.org/doc/apps/public_key/public_key#cacerts_get/0)).
+  SystemCertificates
+  /// Use the system's certificates, along with `cacerts` more, which is a list of
+  /// der-encoded CA certificates as BitArrays. A pem-file cannot be used with this option
+  /// because der-encoded certs have a higher priority over pem-encoded cert files, and the
+  /// `cacerts_get/0` function returns der-encoded certs.
+  WithSystemCertificatesAnd(cacerts: List(BitArray))
+}
+
+pub fn with_ssl(options: ConnectionOptions) -> SSLConnectionOptions {
+  let mug.ConnectionOptions(host, port, timeout) = options
+  SSLConnectionOptions(host, port, timeout, cacerts: SystemCertificates)
+}
+
+pub fn with_cacerts(
+  options: SSLConnectionOptions,
+  cacerts: CaCertificates,
+) -> SSLConnectionOptions {
+  SSLConnectionOptions(..options, cacerts: cacerts)
+}
+
+// TODO: Support certs_keys option of [common_option_cert](https://www.erlang.org/doc/apps/ssl/ssl.html#t:common_option_cert/0).
+fn with_certs_keys() {
+  todo as "not implemented"
+}
+
+/// Establish a TLS-encrypted TCP connection to the server specified in the
+/// connection options.
+///
+/// Returns an error if the connection could not be established.
+///
+/// The socket is created in passive mode, meaning the the `receive` function is
+/// to be called to receive packets from the client. The
+/// `receive_next_packet_as_message` function can be used to switch the socket
+/// to active mode and receive the next packet as an Erlang message.
+///
+pub fn connect(options: SSLConnectionOptions) -> Result(Socket, Error) {
+  let host = charlist.from_string(options.host)
+  use _ <- result.try(ssl_start())
```

Comment:
It's not the job of library code to start OTP applications. Please remove this.

Instead the library could depend on the `ssl` application, or the documentation could tell the user to depend on the `ssl` application.

Reply:
Sorry, I don't understand what you mean. Is the user supposed to start the SSL application themselves? In that case I guess it'd be best to expose a `start()` and `stop()` function. `start()` could also return the default `SslConnectionOptions`, which would mean I'd be able to remove `with_ssl`. I guess I could do it in a function that accepts a callback so that users could use `use` too.

Please let me know your thoughts on this @lpil.

Reply:
Application dependencies are declared in `.app/mix.exs/gleam.toml` and then the init system brings up applications in the required order. Outside of the Erlang shell you should never be starting and stopping applications manually.

Reply:
Ah I see, that's really cool.

I was looking through glisten's source code, and it looks like the library starts the application there:

```gleam
// src/glisten.gleam

/// Start the SSL server with the given handler on the provided port.  The key
/// and cert files must be provided, valid, and readable by the current user.
pub fn serve_ssl(
  handler: Handler(user_message, data),
  port port: Int,
  certfile certfile: String,
  keyfile keyfile: String,
) -> Result(Subject(supervisor.Message), StartError) {
  start_ssl_server(handler, port, certfile, keyfile)
  |> result.map(get_supervisor)
}

pub fn start_ssl_server(
  handler: Handler(user_message, data),
  port port: Int,
  certfile certfile: String,
  keyfile keyfile: String,
) -> Result(Server, StartError) {
  let assert Ok(_nil) = ssl.start()
  // ...
}
```

I was just wondering why glisten doesn't ask the user to depend on the ssl application.

Reply:
Very odd, I don't know why that is.

@rawhat Why is it that Glisten is working outside of the OTP init system here? https://github.com/rawhat/glisten/blob/32c3b97f20d3699ee6b6759b2747246a83286736/src/glisten.gleam#L351

Reply:
I was concerned about it failing to start on certain systems, and this was my not very good solution that I didn't think too much about.

I think it's probably totally reasonable to just add it to `extra_applications` for `glisten`.  I've pushed a change for that.  The next release should be properly conformant 🙏🏻

Reply:
What about the usecases without ssl? That'll result in the ssl application being started unnecessarily.

Reply:
What problem would you foresee arising from that?

Reply:
I'm new to this, so I may be wrong, but if the ssl application is turned on even when the user is not using the TLS features of glisten, and the ssl app crashes for whatever reason, then it'll bring down the entire runtime, right?

Reply:
It would not, no. Erlang's process isolation and supervision system is used to prevent this. The `ssl` application is part of Erlang, it's probably the most trustworthy of all the application.

Reply:
Alright, I'll add `ssl` to the `extra_applications` of mug.